### PR TITLE
feat: allow slim_check to do work in MetaM

### PR DESCRIPTION
**Pull request by kim-em**
This opens the possibility to e.g. run `norm_num` inside a `Testable` instance.

Ideally we would actually add a `Testable` instance that demonstrates this. :-)
Some hurdles:
* `norm_num` can't decide all equalities in the reals. Does it just respond with `success` if it can't decide? We could add an extra case to `TestResult` called `inconclusive`, but I think the code paths would be the same as for `success`.
* alternatively, we could have a new type class `MetaDecidable`, for propositions with decision procedures in `MetaM`. There would be a lot of new design there.
* Until we merge #3114 it's annoying to work on `slim_check`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Control/Random.lean [hunk 1]
```diff
@@ -36,37 +36,41 @@ defining objects that can be created randomly.
...
+`Rand m α` should be thought of a random value in `m α`. -/
+abbrev Rand := RandG StdGen
+
+instance [MonadLift m n] : MonadLiftT (RandG g m) (RandG g n) where
+  monadLift x := fun s => x s
 
 /-- `Random α` gives us machinery to generate values of type `α` -/
 class Random (α : Type u) where
-  random [RandomGen g] : RandG g α
+  random [RandomGen g] : RandG g Id α
```
Comment (eric-wieser) [inline Mathlib/Control/Random.lean:51]: Is there a reason these don't generalize across monads `m`?

File: Mathlib/Control/Random.lean [hunk 2]
```diff
@@ -120,19 +130,26 @@ instance {n : Nat} : BoundedRandom (Fin n) where
...
   let rng := ULift.up stdGen
-  let (res, new) := Id.run <| StateT.run cmd rng
+  let (res, new) ← StateT.run cmd rng
   stdGenRef.set new.down
   pure res
 
-def IO.runRandWith (seed : Nat) (cmd : Rand α) : BaseIO α := do
-  pure $ (cmd.run (ULift.up $ mkStdGen seed)).1
+instance : MonadLift (Rand m) m where
+  monadLift := runRand
```
Comment (eric-wieser) [inline Mathlib/Control/Random.lean:152]: This strikes me as a bit risky; I'd rather not accidentally run a random monad with the default generator and get something less deterministic than expecfted.

**Comment by joneugster** [comment]
@semorrison what's the status of this PR? If you'd still like this to be reviewed, could you resolve the merge conflicts

**Comment by joneugster** [comment]
@semorrison I'm going to close the PR since it has been stale for a while and it's not immediately clear to me if it would be worthwhile for somebody else to adopt it.

Please reopen it anytime, for example as "draft" or the "please-adopt" label if you feel like this would be better!

**Pull request closed without merging**